File: src/cti/schemas/observable.py
import uuid
from datetime import datetime
from pydantic import BaseModel, Field
from cti.models.observable import ObservableType


class ObservableCreate(BaseModel):
    type: ObservableType
    value: str = Field(..., min_length=1, max_length=2048)
    confidence_score: int = Field(default=75, ge=0, le=100)


class ObservableSourceResponse(BaseModel):
    feed_id: uuid.UUID
    feed_name: str
    source_confidence: int
    native_confidence: int
    first_seen_by_feed: datetime
    last_seen_by_feed: datetime

    model_config = {"from_attributes": True}


class ObservableResponse(BaseModel):
    id: uuid.UUID
    type: ObservableType
    value: str
    confidence_score: int
    first_seen: datetime | None
    last_seen: datetime | None
    source_count: int = 0
    sources: list[ObservableSourceResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ObservableListResponse(BaseModel):
    items: list[ObservableResponse]
    total: int
    page: int
    size: int
    pages: int

File: frontend/src/types/observable.ts
export type ObservableType =
  | 'ip-addr'
  | 'domain-name'
  | 'url'
  | 'file-hash'
  | 'email-addr'
  | 'command-line'

export interface ObservableSource {
  feed_id: string
  feed_name: string
  source_confidence: number
  first_seen_by_feed: string
  last_seen_by_feed: string
}

export interface Observable {
  id: string
  type: ObservableType
  value: string
  confidence_score: number
  first_seen: string | null
  last_seen: string | null
  source_count: number
  sources: ObservableSource[]
  created_at: string
  updated_at: string
}

export interface ObservableListResponse {
  items: Observable[]
  total: number
  page: number
  size: number
  pages: number
}

export interface ObservableFilters {
  q?: string
  type?: ObservableType | ''
  confidence_min?: number
  source?: string
  feed_id?: string
  last_seen_after?: string
  page?: number
  size?: number
  sort?: string
  order?: 'asc' | 'desc'
}

export interface ObservableCreate {
  type: ObservableType
  value: string
  confidence_score: number
}
